### ISSUE TITLE
Providing a value of an incorrect type inside a composite makes codec enter an infinite loop

### ISSUE THREAD

**Issue by s0me0ne-unkn0wn** [paritytech/scale-value#47]
Minimal reproducible example:

```rust
use scale_info::{MetaType, PortableRegistry, Registry, TypeInfo};
use scale_value::{scale::encode_as_type, Value};

#[derive(TypeInfo)]
struct A(bool);

fn main() {
    let a = MetaType::new::<A>();
    let mut types = Registry::new();
    let id_a = types.register_type(&a);
    let types: PortableRegistry = types.into();

    let value = Value::unnamed_composite(vec![
        Value::from_bytes(0u32.to_le_bytes())
    ]);

    let mut bytes = Vec::new();
    encode_as_type(&value, id_a.id, &types, &mut bytes).unwrap();
}
```

Gets stuck in this loop forever:

https://github.com/paritytech/scale-value/blob/c5c35123cdcb1748371f1e06d0f339e185b9a523/src/scale_impls/encode.rs#L217-L223

**Comment by jsdw** [comment]
Thanks for the issue!

Ah, yes so if we get an error on `if let Ok(()) = value.encode_as_type_to(type_id.clone(), types, &mut temp_out) { ` we'll loop around and try to dig a layer deeper, but we need to bail out if we can't go any deeper else we'll just keep trying the same wrong type over and over.

**Comment by jsdw** [comment]
The fix for this has been released in 0.15.1
